libsql-replication/src/meta.rs

```diff
@@ -0,0 +1,161 @@
+    InvalidMetaFile,
+    #[error("IO error: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("Invalid log id")]
+    InvalidLogId,
+    #[error("Replica log incompatible with log from handshake")]
+    LogIncompatible,
+    #[error("Failed to commit replication index")]
+    FailedToCommit(std::io::Error),
+}
+
+#[repr(C)]
+#[derive(Debug, Pod, Zeroable, Clone, Copy)]
+pub struct WalIndexMetaData {
+    /// id of the replicated log
+    log_id: u128,
+    /// committed frame index
+    pub committed_frame_no: FrameNo,
+    _padding: u64,
+}
+
+impl WalIndexMetaData {
+    async fn read(file: impl AsyncRead) -> Result<Option<Self>, Error> {
+        pin!(file);
+        let mut buf = [0; size_of::<WalIndexMetaData>()];
+        let meta = match file.read_exact(&mut buf).await {
+            Ok(_) => {
+                let meta: Self =
+                    try_pod_read_unaligned(&buf).map_err(|_| Error::InvalidMetaFile)?;
+                Some(meta)
+            }
+            Err(e) if e.kind() == ErrorKind::UnexpectedEof => None,
+            Err(e) => Err(e)?,
+        };
+
+        Ok(meta)
+    }
+}
+
+pub struct WalIndexMeta {
+    file: File,
+    data: Option<WalIndexMetaData>,
+}
+
+impl WalIndexMeta {
+    pub async fn open(db_path: &Path) -> Result<Self, Error> {
+        let path = db_path.join("client_wal_index");
+
+        tokio::fs::create_dir_all(db_path).await?;
+
+        let mut file = tokio::fs::OpenOptions::new()
+            .create(true)
+            .read(true)
+            .write(true)
+            .open(&path)
+            .await?;
+
+        let data = WalIndexMetaData::read(&mut file).await?;
+
+        Ok(Self { file, data })
+    }
+
+    /// attempts to merge two meta files.
+    pub fn merge_hello(&mut self, hello: HelloResponse) -> Result<(), Error> {
+        let hello_log_id = Uuid::from_str(&hello.log_id)
+            .map_err(|_| Error::InvalidLogId)?
+            .as_u128();
+
+        match self.data {
+            Some(meta) => {
+                if meta.log_id != hello_log_id {
+                    Err(Error::LogIncompatible)
+                } else {
+                    Ok(())
+                }
+            }
+            None => {
+                self.data = Some(WalIndexMetaData {
+                    log_id: hello_log_id,
+                    committed_frame_no: FrameNo::MAX,
+                    _padding: 0,
+                });
+                Ok(())
+            }
+        }
+    }
+
+    pub async fn flush(&mut self) -> std::io::Result<()> {
+        if let Some(data) = self.data {
+            // FIXME: we can save a syscall by calling read_exact_at, but let's use tokio API for now
+            self.file.seek(SeekFrom::Start(0)).await?;
+            let s = self.file.write(bytes_of(&data)).await?;
+            // WalIndexMeta is smaller than a page size, and aligned at the beginning of the file, if
+            // should always be written in a single call
```

Comment:
why not just `write_all`? Makes the code more self-descriptive, and we're future-proof against weird platforms (think some magic wasm/wasi targets)

Reply:
you are correct, I meant to use write_all 👍

Reply:
This was just copied over from sqld, but you are right, write_all makes more sense